Add cypress tests

thrijith opened the pull request:
### Description of the Change

- Setup cypress automated testing

<!-- Enter any applicable Issues here. Example: -->
Closes #91 

### Alternate Designs

N/A

### Possible Drawbacks

N/A

### Verification Process

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [ x My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Changelog Entry

<!--
Add sample CHANGELOG.md entry for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.
Example:
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability
-->

### Credits

<!-- Please list any and all contributors on this PR and any linked issue so that they can be added to this projects CREDITS.md file. -->
Props @

thrijith committed:
Add check before adding event listener

thrijith committed:
Initial setup for Cypress test
Add admin page tests
Add block editor tests

thrijith committed:
Add option to set twitter creds

Update permission of test core version setting

Update description for edit page

Remove test for actual tweet test

Remove secrets passed in test action

thrijith committed:
Update composer to add yoast polyfills library

thrijith committed:
Test tweet feature

Update tests and add  block editor and classic editor cypress tests
Remove wpacceptance

Pass env vars via command line option

jeffpaul committed:
update wp tested-up-to badge in README.md

dependabot[bot] committed:
build(deps): bump ajv from 6.10.2 to 6.12.6

Bumps [ajv](https://github.com/ajv-validator/ajv) from 6.10.2 to 6.12.6.
- [Release notes](https://github.com/ajv-validator/ajv/releases)
- [Commits](https://github.com/ajv-validator/ajv/compare/v6.10.2...v6.12.6)

---
updated-dependencies:
- dependency-name: ajv
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump ajv from 6.10.2 to 6.12.6

Bumps [ajv](https://github.com/ajv-validator/ajv) from 6.10.2 to 6.12.6.
- [Release notes](https://github.com/ajv-validator/ajv/releases)
- [Commits](https://github.com/ajv-validator/ajv/compare/v6.10.2...v6.12.6)

---
updated-dependencies:
- dependency-name: ajv
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Sidsector9 committed:
feat: add colored icons for autoshare status

Sidsector9 committed:
fix: code review improvisations

dependabot[bot] committed:
build(deps): bump tar from 4.4.8 to 4.4.19

Bumps [tar](https://github.com/npm/node-tar) from 4.4.8 to 4.4.19.
- [Release notes](https://github.com/npm/node-tar/releases)
- [Changelog](https://github.com/npm/node-tar/blob/main/CHANGELOG.md)
- [Commits](https://github.com/npm/node-tar/compare/v4.4.8...v4.4.19)

---
updated-dependencies:
- dependency-name: tar
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump tar from 4.4.8 to 4.4.19

Bumps [tar](https://github.com/npm/node-tar) from 4.4.8 to 4.4.19.
- [Release notes](https://github.com/npm/node-tar/releases)
- [Changelog](https://github.com/npm/node-tar/blob/main/CHANGELOG.md)
- [Commits](https://github.com/npm/node-tar/compare/v4.4.8...v4.4.19)

---
updated-dependencies:
- dependency-name: tar
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

thrijith committed:
Add upload artifact action for debugging

Update upload artifact action for debugging

Update upload artifact action for debugging

thrijith committed:
Add wp-env conf

Check if dialog is working in CI

Revert "Check if dialog is working in CI"

This reverts commit 85adb377969ffa4612e81d951096a7e2ed72dd84.

thrijith committed:
Add composer to test run

Add composer to test run fix syntax

Set permalink for unique URL

Add helper function to get random text

Set video to true for testing

Revert "Set video to true for testing"

This reverts commit 1446c60048f69ecd57b2d28f2c6d73dff4c65905.

jeffpaul commented:
@thrijith what else is needed here to bring out of draft?  (also, THANKS for helping on this!)

thrijith commented:
> @thrijith what else is needed here to bring out of draft? (also, THANKS for helping on this!)

Hi @jeffpaul, the last test ran successfully here https://github.com/10up/autoshare-for-twitter/runs/5512545638?check_suite_focus=true#step:8:155 but I am seeing issue with same test when run against different WP version, don't think it is due version but something else, I'll check and see how it goes and update the PR when ready, hope that works 🙂

cadic reviewed (dismissed):
reviewed by accident

thrijith committed:
Move failing tests in matrix test to individual file, update cypress config to enable video

Rename test files for order

Add more time to check for tweet

Try fixing issue with tests in different WP versions

Add timeout before click and additional step for lower version metabox

Reduce timeouts in tests

Temporarily remove trunk from matrix tests

Revert "Temporarily remove trunk from matrix tests"

This reverts commit 1a5576b32f5397b6d5133d5713d60080243db618.

iamdharmesh committed:
Added 'yoast/phpunit-polyfills' lib

iamdharmesh committed:
Removed "wp-acceptance" cases

thrijith commented:
Hi @jeffpaul 

I have the test ready, but there are issues when it is running on GH actions, and the videos generated as part of the process doesn't help in pinpointing the issue here, would be really helpful if someone can take a look at this as well.

- Initially the test were [failing](https://github.com/10up/autoshare-for-twitter/actions/runs/2004947992) for all WP versions, but one passed when I ran it [individually](https://github.com/10up/autoshare-for-twitter/runs/5629860786?check_suite_focus=true).

https://github.com/10up/autoshare-for-twitter/blob/feature/add-cypress-tests/.github/workflows/test.yml#L64-L68

- The [minimum](https://github.com/10up/autoshare-for-twitter/runs/5629860895?check_suite_focus=true) and [trunk](https://github.com/10up/autoshare-for-twitter/runs/5630920530?check_suite_focus=true) versions fail too and the videos generated in test don't have info adding links to the results https://drive.google.com/drive/folders/1WG8SnhQglIgLHvwETy1DBm9BG6QywNi6?usp=sharing

- I ran those test locally and they pass so not sure how to proceed on this one right now, test on [WP minimum](https://cln.sh/7Ql1TF), test on [WP trunk](https://cln.sh/xHaKmt)

iamdharmesh committed:
Moved Update permalink structure to initialize script.

iamdharmesh committed:
Disabled video and  Upload artifacts in case of failure only.

iamdharmesh commented:
Hi @thrijith, 

Test failure in `minimum` and `trunk` env issue is fixed now. You may continue now on this to make it ready for review.

Thanks

thrijith commented:
> Test failure in `minimum` and `trunk` env issue is fixed now. You may continue now on this to make it ready for review.

@iamdharmesh awesome, thanks! also what was the issue though? I see you moved permalink setup to be done after env setup was that is?

I'll check the requirements in https://github.com/10up/autoshare-for-twitter/issues/91and see if anything new needs to be done here or not in sometime.

iamdharmesh commented:
> @iamdharmesh awesome, thanks! also what was the issue though? I see you moved permalink setup to be done after env setup was that is?

@thrijith The issue is related to some file permissions when we specify custom WordPress version for wp-env in .wp-env.json or .wp-env-override.json. In this case .htaccess file couldn’t be created in the root folder of the testing website. This is a known issue and fixed by @cadic. So, I applied his solution here

thrijith committed:
Update lock files

thrijith committed:
Merge branch 'develop' into feature/add-cypress-tests

thrijith committed:
Update lock files

thrijith committed:
Delete post meta if the checkbox for tweet is unchecked, causing it to tweet in classic editor

thrijith committed:
Add e2e test for following cases
1 `A test that checked the autoshare box, saved the draft, unchecked the autoshare box, published the post, and verified no tweet was sent`
2 `A test that saved a draft with the autoshare box unchecked, checked the box on the following pageload, published the post, and verified that the tweet was sent`

thrijith committed:
Remove test from running on push to feature branch

thrijith committed:
Fix issue with element in WP 5.2.11
Move tests to different files for understanding issue in GH action

iamdharmesh commented:
@thrijith considering your comment here https://github.com/10up/autoshare-for-twitter/issues/91#issuecomment-1082345137, making this PR Ready for review.

dinhtungdu reviewed (commented):
@thrijith overall, the PR looks good to me, great work 🎉 !

I notice you use `cy.wait` quite often, can you try to reduce or eliminate the usage of timeout functions? Because using timeout is a signal of potential flaky tests.

iamdharmesh committed:
Merge branch 'develop' of https://github.com/10up/autoshare-for-twitter into feature/add-cypress-tests

iamdharmesh committed:
Added @10up/cypress-wp-utils package and started on improve tests.

iamdharmesh committed:
improved block-editor-a-test.js tests.

iamdharmesh committed:
fix uncheck "enable default"

iamdharmesh committed:
Improve block editor tests.

iamdharmesh committed:
Improve classic editor tests

iamdharmesh committed:
remove unwanted "cypress-file-upload" package.

iamdharmesh committed:
rename test.yml to cypress & update 5.2.11 -> 5.2 WP minimum

iamdharmesh committed:
further improvements to block editor tests.

iamdharmesh committed:
further improvements to classic editor tests.

iamdharmesh committed:
Fixed WP minimum failures.

iamdharmesh commented:
Hi @dinhtungdu,

I have updated PR for utilizing our utils lib **@10up/cypress-wp-utils** and improved tests by removing wait and timeout-related stuff.

Please help to re-review it.

Thanks
cc: @thrijith

dinhtungdu commented on tests/cypress/support/commands.js:
```diff
@@ -0,0 +1,55 @@
+import '@10up/cypress-wp-utils';
+import { getRandomText } from "../support/functions";
+
+Cypress.Commands.add( 'visitAdminPage', ( page = 'index.php' ) => {
+	cy.login();
+	if ( page.includes( 'http' ) ) {
+		cy.visit( page );
+	} else {
+		cy.visit( `/wp-admin/${ page.replace( /^\/|\/$/g, '' ) }` );
+	}
+});
```
We have a conversation [here](https://github.com/10up/cypress-wp-utils/pull/1) about not adding the `visitAdminPage` command. I think we should follow the conclusion there in our plugin as well. What do you think? Also cc @cadic.

iamdharmesh replied:
Done.

iamdharmesh committed:
Removed visitAdminPage command.

dinhtungdu commented on tests/cypress/integration/block-editor.test.js:
```diff
@@ -8,16 +8,20 @@ describe('Test Autoshare for Twitter with Block Editor.', () => {
 		});
 	});
 
+	beforeEach(() => {
+		cy.login();
+	});
```
What do you think about applying [this technique](https://github.com/10up/cypress-wp-utils/pull/46) to improve the testing performance?

iamdharmesh replied:
Yes, it would be great to apply it. but it's not merged yet.

iamdharmesh replied:
Ohh. you are asking to use that technique directly here on this plugin. Yes, let me do that. I was waiting for get it merged on our utils lib. :D

dinhtungdu replied:
I think we just need to copy [these lines](https://github.com/10up/cypress-wp-utils/pull/46/files#diff-6bd80aef2afe0d5b3dcd8c0f268a82b3d0b197ef23fab87ee54145dd12efe09a) and update the tests.

dinhtungdu replied:
I did [the same ](https://github.com/10up/simple-podcasting/pull/162/files#diff-6bd80aef2afe0d5b3dcd8c0f268a82b3d0b197ef23fab87ee54145dd12efe09a)for simple-podcasting.

dinhtungdu reviewed (dismissed):
LGTM! Left an optional comment but pre-approving this! Thanks so much for the great work on this PR!

iamdharmesh committed:
Preserve WP cookies for one time login

iamdharmesh committed:
fix typo.
